src/main/java/com/siit/hospital_manager/controller/AppointmentController.java

```diff
@@ -21,27 +27,138 @@ public class AppointmentController {
 
     private final AppointmentService appointmentService;
     private final DoctorService doctorService;
+    private final PatientService patientService;
+    private final SpecialtyRepository specialtyRepository;
+    private final DiagnosisService diagnosisService;
+    private final ProcedureService procedureService;
+    private final MedicationService medicationService;
 
-    @GetMapping("/findAllByPatient")
+    private final EmailSender emailSender;
+
+    @GetMapping("/viewAll")
     public String findAllByPatient(Model model, Principal principal) {
         List<AppointmentDto> appointments = appointmentService.findAllByUserName(principal.getName());
         model.addAttribute("appointments", appointments);
-
         return "appointment/viewAll";
     }
 
+    @GetMapping("/appointmentDetails/{apId}")
+    public String findAllDetailsByPatient(@PathVariable Integer apId,Model model) {
+        model.addAttribute("appointment", appointmentService.findById(apId));
+        model.addAttribute("appointmentDiagnoses", appointmentService.findById(apId).getDiagnoses());
+        model.addAttribute("appointmentProcedures", appointmentService.findById(apId).getProcedures());
+        model.addAttribute("appointmentMedications", appointmentService.findById(apId).getMedications());
+        return "appointment/appointmentDetails/appointmentsDetails";
+    }
+
+
+
     @DeleteMapping("/{id}")
     @ResponseStatus(HttpStatus.ACCEPTED)
     public void deleteAppointmentById(Model model, @PathVariable Integer id, Principal principal){
          appointmentService.deleteAppointmentByIdAndPatient(id, principal.getName());
     }
 
-    @GetMapping("/create")
-    public String create(Model model, Principal principal) {
-        model.addAttribute("doctors", doctorService.findAll());
+    @GetMapping("/create/{specialtyId}")
+    public String showCreateSpecialtyForm(@PathVariable Integer specialtyId, Model model) {
+        Specialty specialty = specialtyRepository.findById(specialtyId)
+                .orElseThrow(() -> new EntityNotFoundException("Specialty not found"));
+        model.addAttribute("doctors", doctorService.findAllBySpecialty(specialty));
+        model.addAttribute("appointment", CreateAppointmentDto.builder().build());
+        return "appointment/createAppointment";
+    }
+
+    @PostMapping("/submitCreateAppointmentForm")
+    public String submitCreateAppointmentForm(@Valid CreateAppointmentDto createAppointmentDto, BindingResult bindingResult, Authentication authentication) {
+        MyUserDetails myUserDetails = (MyUserDetails) authentication.getPrincipal();
+        String username = myUserDetails.getUsername();
+        Patient patient = patientService.findByUsername(username);
+        String userEmail = patient != null ? patient.getEmail() : null;
```

Comment:
you are sending a email to a null email address here if it's not found on the patient, also you should not do work directly in the controller, make a service method

Reply:
use controllers to route between views and treat exceptions from services (if not caught by global exception handler)